tests/nn_test.py

```diff
@@ -36,8 +38,99 @@
+    return False
+  if cuda_versions is None:
+    return False
+  if not jtu.is_cuda_compute_capability_at_least("8.0"):
+    return False
+  if cuda_versions.cudnn_get_version() < 8904:
+    return False
+
+  return True
+
+def _get_causal_mask(T, S):
+  assert T == S
+  mask_shape = (T, S)
+  row_idx = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 0)
+  col_idx = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 1)
+  return (row_idx >= col_idx)[jnp.newaxis, jnp.newaxis, :, :]
+
+@jtu.with_config(jax_legacy_prng_key="allow",
+                 jax_numpy_dtype_promotion="standard")
 class NNFunctionsTest(jtu.JaxTestCase):
+  @parameterized.product(
+      dtype=[jnp.bfloat16, jnp.float16],
+      use_bias=(False, True),
+      causal_mode=(None, 'is_causal', 'is_mask'),
+      impl=('xla', 'cudnn'),
+  )
+  @jtu.skip_on_flag("jax_skip_slow_tests", True)
+  def testDotProductAttentionInfer(self, dtype, use_bias, causal_mode, impl):
+    if impl == 'cudnn' and not _is_required_cudnn_version_satisfied():
+      raise unittest.SkipTest("CUDA or cuDNN versions are not compatible.")
+
+    sdpa = nn.dot_product_attention
+    sdpa_ref = partial(sdpa, implementation=None)
+    sdpa_ans = partial(sdpa, implementation=impl)
+
+    B, S, T, N, H = 4, 1024, 1024, 4, 64
+    keys = random.split(random.PRNGKey(0), 4)
+    Q = random.normal(keys[0], (B, T, N, H), dtype)
+    K = random.normal(keys[1], (B, S, N, H), dtype)
+    V = random.normal(keys[2], (B, S, N, H), dtype)
+    if use_bias:
+      bias = random.normal(keys[3], (1, N, T, S), dtype)
+    else:
+      bias = None
+
+    is_causal = causal_mode == 'is_causal'
+    causal_mask = _get_causal_mask(T, S) if causal_mode == 'is_mask' else None
+
+    out_ref = sdpa_ref(Q, K, V, bias, causal_mask, is_causal=is_causal)
+    out_ans = sdpa_ans(Q, K, V, bias, causal_mask, is_causal=is_causal)
+    self.assertAllClose(out_ref, out_ans, atol=.01, rtol=.01)
+
+  @parameterized.product(
+      dtype=[jnp.bfloat16, jnp.float16],
+      use_bias=(False, True),
+      causal_mode=(None, 'is_causal', 'is_mask'),
+      impl=('xla', 'cudnn'),
+  )
+  @jtu.skip_on_flag("jax_skip_slow_tests", True)
+  def testDotProductAttentionTrain(self, dtype, use_bias, causal_mode, impl):
+    if impl == 'cudnn' and not _is_required_cudnn_version_satisfied():
+      raise unittest.SkipTest("CUDA or cuDNN versions are not compatible.")
+
+    sdpa = nn.dot_product_attention
+    B, S, T, N, H = 4, 1024, 1024, 4, 64
+    keys = random.split(random.PRNGKey(0), 5)
+    Q = random.normal(keys[0], (B, T, N, H), dtype)
+    K = random.normal(keys[1], (B, S, N, H), dtype)
+    V = random.normal(keys[2], (B, S, N, H), dtype)
+    grad = random.normal(keys[3], (B, T, N, H), dtype)
+    if use_bias:
+      bias = random.normal(keys[4], (1, N, T, S), dtype)
+    else:
+      bias = None
+
+    is_causal = causal_mode == 'is_causal'
+    causal_mask = _get_causal_mask(T, S) if causal_mode == 'is_mask' else None
+
```

Comment:
A very important test is to verify that if the user states cudnn as the implementation, this is actually used. I think we should check the HLO, which you can get like:
```
jax.jit(f).lower(args).compiler_ir('stablehlo')
```

Reply:
Done.